src/AggregateFunctions/SingleValueData.h

```diff
@@ -0,0 +1,393 @@
+    bool isEqualTo(const IColumn & column, size_t index) const override;
+    bool isEqualTo(const SingleValueDataBase & to) const override;
+
+    void set(const IColumn & column, size_t row_num, Arena * arena) override;
+    void set(const SingleValueDataBase & to, Arena * arena) override;
+
+    bool setIfSmaller(const SingleValueDataBase & to, Arena * arena) override;
+    bool setIfGreater(const SingleValueDataBase & to, Arena * arena) override;
+    bool setIfSmaller(const IColumn & column, size_t row_num, Arena * arena) override;
+    bool setIfGreater(const IColumn & column, size_t row_num, Arena * arena) override;
+    void setSmallest(const IColumn & column, size_t row_begin, size_t row_end, Arena * arena) override;
+    void setGreatest(const IColumn & column, size_t row_begin, size_t row_end, Arena * arena) override;
+    void setSmallestNotNullIf(
+        const IColumn & column,
+        const UInt8 * __restrict null_map,
+        const UInt8 * __restrict if_map,
+        size_t row_begin,
+        size_t row_end,
+        Arena * arena) override;
+    void setGreatestNotNullIf(
+        const IColumn & column,
+        const UInt8 * __restrict null_map,
+        const UInt8 * __restrict if_map,
+        size_t row_begin,
+        size_t row_end,
+        Arena * arena) override;
+
+    std::optional<size_t> getSmallestIndex(const IColumn & column, size_t row_begin, size_t row_end) const override;
+    std::optional<size_t> getGreatestIndex(const IColumn & column, size_t row_begin, size_t row_end) const override;
+    std::optional<size_t> getSmallestIndexNotNullIf(
+        const IColumn & column,
+        const UInt8 * __restrict null_map,
+        const UInt8 * __restrict if_map,
+        size_t row_begin,
+        size_t row_end) const override;
+    std::optional<size_t> getGreatestIndexNotNullIf(
+        const IColumn & column,
+        const UInt8 * __restrict null_map,
+        const UInt8 * __restrict if_map,
+        size_t row_begin,
+        size_t row_end) const override;
+
+    static bool allocatesMemoryInArena() { return false; }
+};
+
+#define DISPATCH(TYPE) \
+    extern template struct SingleValueDataNumeric<TYPE>; \
+    static_assert( \
+        sizeof(SingleValueDataNumeric<TYPE>) <= SingleValueDataBase::MAX_STORAGE_SIZE, "Incorrect size of SingleValueDataNumeric struct");
+
+FOR_SINGLE_VALUE_NUMERIC_TYPES(DISPATCH)
+#undef DISPATCH
+
+
+/** For strings. Short strings are stored in the object itself, and long strings are allocated separately.
+  * NOTE It could also be suitable for arrays of numbers.
+//  */
+struct SingleValueDataString final : public SingleValueDataBase
+{
+    static constexpr bool is_compilable = false;
+    using Self = SingleValueDataString;
+
+    /// 0 size indicates that there is no value. Empty string must has terminating '\0' and, therefore, size of empty string is 1
+    UInt32 size = 0;
+    UInt32 capacity = 0; /// power of two or zero
+    char * large_data;
```

Comment:
where do we free it?

Reply:
It's allocated in an arena, so it's freed with the arena. Adding a comment